<a href="https://colab.research.google.com/github/djmary-k/Py16DS_HW03/blob/main/Hw3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Homework 03, Module 03, Data Science.
Дане домашнє завдання буде повністю пов'язане з лінійною регресією та її реалізацією.

Розіб'ємо наше домашнє завдання на декілька частин:
* напишіть функцію гіпотези лінійної регресії у векторному вигляді;
* створіть функцію для обчислення функції втрат у векторному вигляді;
* реалізуйте один крок градієнтного спуску;

In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

In [5]:
def cost(h, y):
  return np.mean((h-y)**2)/2

class CustomLinearRegression:
  def __init__(self, lr: float = 0.01, thr=0.000000001, n_epochs = 1000):
    # lr - learning rate(коефіцієнт навчання, який відповідає за те, наскільки великий крок буде робити алгоритм на кожній ітерації)
    # thr - критерій зупинки навчання або трешхолд
    # n_epoch - кількість епох\ кроки градієнтного спуску (кількість ітерацій \ num_iter)
    self.lr = lr # alfa
    self.thr = thr # epsilon
    self.n_epochs = n_epochs # num_iter
    # self.w = np.array([np.random.uniform(-5,5)])
    self.w = np.array([np.random.uniform(-5,5), np.random.uniform(-5,5), np.random.uniform(-5,5)])

  def predict(self, X): # гіпотетична модель h(x)
    return self.w @ X.T

  def update_weights(self, X, y): #реалізація одиного кроку градієнтного спуску grad_step
    m = len(y)
    h = self.predict(X)
    self.w = self.w - self.lr / m * X.T @ (h-y)

  def fit(self, X, y): # цикл із кроків градієнтного спуску grade_descent
    last_cost = 10000000 #рандомне величезне число
    for i in range(self.n_epochs):
      self.update_weights(X, y)
      new_cost = cost(self.predict(X), y)

      if i % (self.n_epochs//1000) == 0:
        print(f'cost function = {last_cost}')

      if abs(last_cost - new_cost) < self.thr: # зупинка навчання
        break

      last_cost = new_cost



* **знайдіть найкращі параметри $w$ для датасету прогнозуючу ціну на будинок залежно від площі, кількості ванних кімнат та кількості спалень**

In [6]:
data_housing = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Housing.csv')
data_housing.head()

price  area  bedrooms  bathrooms  stories mainroad guestroom basement  \
0  13300000  7420         4          2        3      yes        no       no   
1  12250000  8960         4          4        4      yes        no       no   
2  12250000  9960         3          2        2      yes        no      yes   
3  12215000  7500         4          2        2      yes        no      yes   
4  11410000  7420         4          1        2      yes       yes      yes   

  hotwaterheating airconditioning  parking prefarea furnishingstatus  
0              no             yes        2      yes        furnished  
1              no             yes        3       no        furnished  
2              no              no        2      yes   semi-furnished  
3              no             yes        3      yes        furnished  
4              no             yes        2       no        furnished

In [ ]:
# data_housing.plot(x='area', y='price', kind='scatter')

In [7]:
# Split the data into training/testing sets
X_train = data_housing.loc[:500, ['area', 'bathrooms', 'bedrooms']]
X_test = data_housing.loc[501:, ['area', 'bathrooms', 'bedrooms']]

# Split the targets into training/testing sets
y_train = data_housing.loc[:500, 'price']
y_test = data_housing.loc[501:, 'price']



In [8]:
# normalization data: 'X' and 'y'

def normalization(data):
  mean_data = np.mean(data)
  value_range = np.max(data) - np.min(data)
  result = []

  for x in data:
    norm_x = (x - mean_data) / value_range
    result.append(norm_x)

  return result

In [ ]:
# # y - Vector [prices]
# # X - Matrix [area]

# norm_y = np.array(normalization(data_housing.price))
# norm_x = pd.DataFrame()
# norm_x['area'] = normalization(data_housing.area)

In [9]:
# y - Vector [prices]
# X - Matrix [area, bathrooms, bedrooms]

norm_y = np.array(normalization(y_train))
norm_x = pd.DataFrame()
norm_x['area'] = normalization(X_train.area)
norm_x['bathrooms'] = normalization(X_train.bathrooms)
norm_x['bedrooms'] = normalization(X_train.bedrooms)

In [ ]:
cust_regression = CustomLinearRegression(lr=0.1, thr=0.00000000000001)
# cust_regression = CustomLinearRegression(lr=0.1, thr=0.000001)
cust_regression.fit(norm_x, norm_y)

In [24]:
cust_regression.w

area         0.223417
bathrooms    0.502759
bedrooms     0.357496
dtype: float64

* **знайдіть ці ж параметри за допомогою аналітичного рішення**

In [11]:
# 1
def analytical_solution(X, y):
    return np.linalg.pinv(X) @ y

In [28]:
w_analytical_1 = analytical_solution(norm_x, norm_y)
w_analytical_1

array([0.46607398, 0.37812754, 0.17297334])

In [43]:
# 2
w_analytical_2 = np.linalg.inv(norm_x.T @ norm_x) @ norm_x.T @ norm_y
w_analytical_2

0    0.466074
1    0.378128
2    0.172973
dtype: float64

* **реалізація за допомогою бібліотеки sklearn**

In [31]:
from sklearn.linear_model import LinearRegression
regressor = LinearRegression().fit(norm_x, norm_y)
# коефіцієнти
print(regressor.coef_)
# зсув
print(regressor.intercept_)

[0.46607398 0.37812754 0.17297334]
-2.95177287828457e-17


* **порівняйте отримані результати**

Отже, параметри $w$ для даного датасету:
1. Отримані за допомогою кастомної реалізації

**area         0.223417**

**bathrooms    0.502759**

**bedrooms     0.357496**

2. Отримані за допомогою аналітичних рішеннь:
- **[0.46607398, 0.37812754, 0.17297334]**
- **0    0.466074** / **1    0.378128** / **2    0.172973**

3. Отримані за допомогою бібліотеки scikit learn:

**[0.47714269 0.36001286 0.17611257]**


*Щоб визначити який результат кращий, спрогнозуємо ціни для датасету **X_test** за допомогою отриманих параметрів $w$ і порівняємо їх з цільовим датасетом **y_test**.*

In [70]:
# Make predictions using the testing set
# 1 solution
cust_solution = cust_regression.predict(X_test)
# 2 solution
analytical_solution_1 = X_test @ w_analytical_1
analytical_solution_2 = X_test @ np.array(w_analytical_2)
# 3 solution
sklearn_solution = regressor.predict(X_test)

In [69]:
result = pd.DataFrame(
    {
        'Target prices': y_test,
        'Prediction prices (custom)': cust_solution,
        'Prediction prices (analytical 1)': analytical_solution_1,
        'Prediction prices (analytical 2)': analytical_solution_2,
        'Prediction prices (sklearn)': sklearn_solution
    }
)
result

Target prices  Prediction prices (custom)  \
501        2660000                  544.477515   
502        2660000                  778.707419   
503        2660000                  895.241532   
504        2653000                  712.799530   
505        2653000                  895.241532   
506        2604000                  651.359973   
507        2590000                  805.517408   
508        2590000                  984.250665   
509        2590000                  806.020167   
510        2520000                  645.014972   
511        2520000                  712.039944   
512        2520000                  671.467465   
513        2485000                  984.608161   
514        2485000                  671.824961   
515        2450000                  718.742441   
516        2450000                  725.087442   
517        2450000                  671.467465   
518        2450000                  783.175750   
519        2450000                 1082.553956   
520        2450000                 1721.525350   
521        2408000                  813.336988   
522        2380000                  554.531261   
523        2380000                  625.097486   
524        2380000                  730.449440   
525        2345000                  814.454070   
526        2310000                  711.682448   
527        2275000                  411.410576   
528        2275000                  887.824043   
529        2275000                  888.539035   
530        2240000                  437.237561   
531        2233000                 1185.683075   
532        2135000                  671.467465   
533        2100000                  537.775018   
534        2100000                  672.182457   
535        2100000                  751.897430   
536        1960000                  766.374912   
537        1890000                  381.383418   
538        1890000                  816.464820   
539        1855000                  669.233299   
540        1820000                  671.467465   
541        1767150                  537.775018   
542        1750000                  809.985739   
543        1750000                  651.717469   
544        1750000                  861.729046   

     Prediction prices (analytical 1)  Prediction prices (analytical 2)  \
501                       1133.456823                       1133.456823   
502                       1622.661530                       1622.661530   
503                       1865.192974                       1865.192974   
504                       1485.169706                       1485.169706   
505                       1865.192974                       1865.192974   
506                       1356.999361                       1356.999361   
507                       1678.590408                       1678.590408   
508                       2051.449593                       2051.449593   
509                       1678.968535                       1678.968535   
510                       1343.190115                       1343.190115   
511                       1483.012309                       1483.012309   
512                       1398.946019                       1398.946019   
513                       2051.622567                       2051.622567   
514                       1399.118992                       1399.118992   
515                       1496.994528                       1496.994528   
516                       1510.803775                       1510.803775   
517                       1398.946019                       1398.946019   
518                       1631.983010                       1631.983010   
519                       2256.522145                       2256.522145   
520                       3589.493732                       3589.493732   
521                       1694.902997                       1694.902997   
522                       1154.430152                       1154.430152   
523                       130